File: javaproject/app/src/main/java/com/marcosvbras/robomarket/business/api/RequestInterceptor.java
package com.marcosvbras.robomarket.business.api;

import com.marcosvbras.robomarket.app.RoboApplication;
import com.marcosvbras.robomarket.utils.Constants;

import java.io.IOException;
import java.util.HashMap;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

class RequestInterceptor implements Interceptor {

    @Override
    public Response intercept(Chain chain) throws IOException {
        // Setting Back4App Credentials
        Request request = chain.request().newBuilder()
                .header(Constants.Api.HEADER_APP_ID, Constants.Api.APP_ID)
                .header(Constants.Api.HEADER_REST_API_KEY, Constants.Api.REST_API_KEY)
                .method(chain.request().method(), chain.request().body())
                .build();

        HashMap<String, String> credentials = RoboApplication.getInstance().getUserCredentials();
        String sessionToken = credentials.get(Constants.Preferences.SESSION_TOKEN_KEY);

        // Setting user session token if it exists
        if(sessionToken != null) {
            request = request.newBuilder()
                    .header(Constants.Api.HEADER_SESSION_TOKEN, sessionToken)
                    .method(request.method(), request.body())
                    .build();
        }

        return chain.proceed(request);
    }
}

File: javaproject/app/src/main/java/com/marcosvbras/robomarket/business/api/RequestInterceptor.kt
package com.marcosvbras.robomarket.business.api

import com.marcosvbras.robomarket.app.App
import com.marcosvbras.robomarket.app.SESSION_TOKEN_PREF
import okhttp3.Interceptor
import okhttp3.Response

class RequestInterceptor : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        var request = chain.request()?.newBuilder()
                ?.header(APIService.HEADER_APP_ID, APIService.APP_ID)
                ?.header(APIService.HEADER_REST_API_KEY, APIService.REST_API_KEY)
                ?.method(chain.request().method(), chain.request()?.body())
                ?.build()

        val credentials = App.instance.userCredentials
        val sessionToken = credentials[SESSION_TOKEN_PREF]

        if(sessionToken != null) {
            request = request?.newBuilder()
                    ?.header(APIService.HEADER_SESSION_TOKEN, sessionToken)
                    ?.method(request.method(), request.body())
                    ?.build()
        }

        return chain.proceed(request)
    }
}